Shared/Managers/KeyboardManager.swift

```diff
@@ -37,88 +41,110 @@ class KeyboardManager {
             .mainSink { (notification) in
-                self.currentEvent = .willShow(notification)
-                self.cachedKeyboardEndFrame = self.getFrame(for: notification)
-
-                if let inputView = self.inputAccessoryView {
-                    let willShow = self.cachedKeyboardEndFrame.height > inputView.height
-                    self.willKeyboardShow = willShow
-                } else {
-                    self.willKeyboardShow = true
+                guard !self.isKeyboardShowing else { return }
+
+                let inputAccessoryHeight = self.getInputAccessoryHeight()
+
+                if notification.keyboardEndFrame.height > inputAccessoryHeight {
+                    self.currentEvent = .willShow(notification)
+                    self.cachedKeyboardEndFrame = notification.keyboardEndFrame
                 }
             }.store(in: &self.cancellables)
 
         NotificationCenter.default.publisher(for: UIResponder.keyboardDidShowNotification)
             .mainSink { (notification) in
-                self.currentEvent = .didShow(notification)
-                self.cachedKeyboardEndFrame = self.getFrame(for: notification)
+                guard !self.isKeyboardShowing else { return }
 
-                if let inputView = self.inputAccessoryView {
-                    let isShowing = self.cachedKeyboardEndFrame.height > inputView.height
-                    self.isKeyboardShowing = isShowing
-                } else {
+                let inputAccessoryHeight = self.getInputAccessoryHeight()
+
+                if notification.keyboardEndFrame.height > inputAccessoryHeight {
+                    self.currentEvent = .didShow(notification)
+                    self.cachedKeyboardEndFrame = notification.keyboardEndFrame
                     self.isKeyboardShowing = true
                 }
-
             }.store(in: &self.cancellables)
 
         NotificationCenter.default.publisher(for: UIResponder.keyboardWillHideNotification)
             .mainSink { (notification) in
-                self.currentEvent = .willHide(notification)
-                self.cachedKeyboardEndFrame = self.getFrame(for: notification)
-                if let inputView = self.inputAccessoryView {
-                    let willShow = self.cachedKeyboardEndFrame.height > inputView.height
-                    self.willKeyboardShow = willShow
-                } else {
-                    self.willKeyboardShow = true
+                guard self.isKeyboardShowing else { return }
+
+                let inputAccessoryHeight = self.getInputAccessoryHeight()
+                let keyboardFrameVisibleHeight = self.getKeyboardFrameVisibleHeight(for: notification)
+
+                if keyboardFrameVisibleHeight <= inputAccessoryHeight {
+                    self.currentEvent = .willHide(notification)
+                    self.cachedKeyboardEndFrame = notification.keyboardEndFrame
```

Comment:
Similar to the "show" scenario, we make sure that the keyboard is actually hiding by comparing the visible portion of the keyboard frame against the input accessory.